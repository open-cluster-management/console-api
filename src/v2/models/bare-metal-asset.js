/** *****************************************************************************
 * Licensed Materials - Property of Red Hat, Inc.
 * (c) Copyright Red Hat, Inc. All Rights Reserved.
 ****************************************************************************** */
import KubeModel from './kube';

function transform(bareMetalAsset, secret = {}) {
  const { spec } = bareMetalAsset;
  const username = secret.data ? secret.data.username : undefined;
  const password = secret.data ? secret.data.password : undefined;

  const bma = {
    metadata: bareMetalAsset.metadata,
    ...spec,
  };

  bma.bmc = bma.bmc || {};
  bma.bmc.username = username ? Buffer.from(username, 'base64').toString('ascii') : undefined;
  bma.bmc.password = password ? Buffer.from(password, 'base64').toString('ascii') : undefined;
  return bma;
}

function transformNamespaces(namespace) {
  return namespace.metadata.name;
}

export default class BareMetalAssetModel extends KubeModel {
  async getSingleBareMetalAsset(args = {}) {
    const { name, namespace } = args;
    if (name && namespace) {
      const bareMetalAsset = await this.kubeConnector.get(`/apis/midas.io/v1alpha1/namespaces/${namespace}/baremetalassets/${name}`);
      let secret;
      if (bareMetalAsset.metadata !== undefined) {
        if (bareMetalAsset.spec.bmc && bareMetalAsset.spec.bmc.credentialsName) {
          secret = await this.kubeConnector.get(`/api/v1/namespaces/${namespace}/secrets/${bareMetalAsset.spec.bmc.credentialsName}`);
        }
        return [transform(bareMetalAsset, secret)];
      }
    }
    return [];
  }

  async getBareMetalAssets(args = {}) {
    const { name } = args;
    const allBareMetalAssets = await this.getAllBareMetalAssets(args);
    if (name) {
      return allBareMetalAssets.find(bma => bma.metadata.name === name);
    }
    return allBareMetalAssets;
  }

  async getAllBareMetalAssets() {
    const [bareMetalAssets] = await Promise.all([
      this.kubeConnector.get('/apis/midas.io/v1alpha1/baremetalassets'),
      // TODO: do we need secrets?
    ]);
    if (bareMetalAssets.items !== undefined) {
      return bareMetalAssets.items.map(bma => transform(bma));
    }
    return [];
  }

  async getBareMetalAssetSubresources(args = {}) {
    const [namespaces, bareMetalAsset] = await Promise.all([
      this.kubeConnector.get('/api/v1/namespaces'),
      this.getSingleBareMetalAsset(args),
    ]);
    return {
      namespaces: namespaces.items ?
        namespaces.items.map(transformNamespaces)
        : { error: namespaces },
      bareMetalAsset,
    };
  }

  async createSecret(namespace, secretName, username, password) {
    const secret = {
      apiVersion: 'v1',
      kind: 'Secret',
      metadata: {
        name: secretName,
      },
      type: 'Opaque',
      data: {
        username: Buffer.from(username).toString('base64'),
        password: Buffer.from(password).toString('base64'),
      },
    };
    return this.kubeConnector.post(`/api/v1/namespaces/${namespace}/secrets`, secret);
  }

  async createBMA(namespace, name, address, credentialsName, bootMACAddress) {
    const bma = {
      apiVersion: 'midas.io/v1alpha1',
      kind: 'BareMetalAsset',
      metadata: {
        name,
      },
      spec: {
        bmc: {
          address,
          credentialsName,
        },
        bootMACAddress,
      },
    };
    return this.kubeConnector.post(`/apis/midas.io/v1alpha1/namespaces/${namespace}/baremetalassets`, bma);
  }

  async patchSecret(namespace, secretName, username, password) {
    const secretBody = {
      body: [
        {
          op: 'replace',
          path: '/data',
          value: {
            username: Buffer.from(username).toString('base64'),
            password: Buffer.from(password).toString('base64'),
          },
        },
      ],
    };
    return this.kubeConnector.patch(`/api/v1/namespaces/${namespace}/secrets/${secretName}`, secretBody);
  }

  async patchBMA(oldSpec, namespace, name, address, credentialsName, bootMACAddress) {
    const newSpec = Object.assign({}, oldSpec, {
      bmc: {
        address,
        credentialsName,
      },
      bootMACAddress,
    });
    const bmaBody = {
      body: [
        {
          op: 'replace',
          path: '/spec',
          value: newSpec,
        },
      ],
    };
    return this.kubeConnector.patch(`/apis/midas.io/v1alpha1/namespaces/${namespace}/baremetalassets/${name}`, bmaBody);
  }

  async createBareMetalAsset(args) {
    const {
      namespace, name, bmcAddress, username, password, bootMac,
    } = args;
    try {
      const secretName = `${name}-bmc-secret`;
      const secretResult = await this.createSecret(namespace, secretName, username, password);
      const bmaResult = await this.createBMA(namespace, name, bmcAddress, secretName, bootMac);

      let statusCode = 201;
      if (secretResult.metadata.name !== secretName) {
        statusCode = secretResult.code;
      } else if (bmaResult.metadata.name !== name) {
        statusCode = bmaResult.code;
      }

      return {
        statusCode,
        secretResult,
        bmaResult,
      };
    } catch (error) {
      return error;
    }
  }

  async updateBareMetalAsset(args) {
    const {
      namespace, name, bmcAddress, username, password, bootMac,
    } = args;

    const bareMetalAsset = await this.kubeConnector.get(`/apis/midas.io/v1alpha1/namespaces/${namespace}/baremetalassets/${name}`);
    if (bareMetalAsset.metadata !== undefined) {
      let secretResult;
      if (bareMetalAsset.spec.bmc && bareMetalAsset.spec.bmc.credentialsName) {
        secretResult = await this.patchSecret(
          namespace,
          bareMetalAsset.spec.bmc.credentialsName,
          username,
          password,
        );
      } else {
        // Secret does not exist, create one
        secretResult = await this.createSecret(namespace, `${name}-bmc-secret`, username, password);
      }
      if (secretResult && (secretResult.code || secretResult.message)) {
        return {
          statusCode: secretResult.code || 500,
          secretResult,
          bmaResult: {},
        };
      }

      const bmaResult = await this.patchBMA(
        bareMetalAsset.spec,
        namespace,
        name,
        bmcAddress,
        secretResult.metadata.name,
        bootMac,
      );
      if (bmaResult && (bmaResult.code || bmaResult.message)) {
        return {
          statusCode: bmaResult.code || 500,
          secretResult,
          bmaResult,
        };
      }

      return {
        statusCode: 200,
        secretResult,
        bmaResult,
      };
    }

    // BMA not found
    return {
      statusCode: 400,
      secretResult: {},
      bmaResult: {},
    };
  }
}
