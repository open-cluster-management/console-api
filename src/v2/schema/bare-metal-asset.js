
/** *****************************************************************************
 * Licensed Materials - Property of Red Hat, Inc.
 * Copyright (c) 2020 Red Hat, Inc.
 ****************************************************************************** */
export const typeDef = `
type BmcType {
  address: String
  credentialsName: String
  username: String
  password: String
}

type BmcClusterDeploymentType {
  name: String
  namespace: String
}

type BareMetalAsset implements K8sObject {
  metadata: Metadata
  bmc: BmcType
  bootMACAddress: String
  clusterDeployment: BmcClusterDeploymentType
  hardwareProfile: String
  role: String
  status: String
}

type BareMetalAssetSubresources {
  namespaces: [String],
  bareMetalAsset: [BareMetalAsset]
}
`;

export const resolver = {
  Query: {
    bareMetalAssets: (root, args, { bareMetalAssetModel }) =>
      bareMetalAssetModel.getBareMetalAssets({ ...args }),
    bareMetalAsset: (root, args, { bareMetalAssetModel }) =>
      bareMetalAssetModel.getSingleBareMetalAsset({ ...args }),
    bareMetalAssetSubresources: (root, args, { bareMetalAssetModel }) =>
      bareMetalAssetModel.getBareMetalAssetSubresources({ ...args }),
  },
  Mutation: {
    createBareMetalAsset: (parent, args, { bareMetalAssetModel }) =>
      bareMetalAssetModel.createBareMetalAsset(args),
    updateBareMetalAsset: (parent, args, { bareMetalAssetModel }) =>
      bareMetalAssetModel.updateBareMetalAsset(args),
    deleteBareMetalAssets: (parent, args, { bareMetalAssetModel }) =>
      bareMetalAssetModel.deleteBareMetalAssets(args),
  },

};
