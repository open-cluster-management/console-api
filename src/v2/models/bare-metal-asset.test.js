/** *****************************************************************************
 * Licensed Materials - Property of Red Hat, Inc.
 * (c) Copyright Red Hat, Inc. All Rights Reserved.
 ****************************************************************************** */
import { transform } from './bare-metal-asset';

const bma1 = {
  metadata: {
    name: 'foo',
    namespace: 'bar',
  },
  spec: {
    bmc: {
      address: 'bmcAddr1',
      credentialsName: 'my-secret-name',
    },
    bootMACAddress: 'addr1',
    clusterDeployment: 'cluster',
    hardwareProfile: 'my-profile',
    role: 'role',
  },
};

const secret1 = {
  metadata: {
    namespace: 'bar',
    name: 'my-secret-name',
  },
  data: {
    username: Buffer.from('myuser1').toString('base64'),
    password: Buffer.from('mypwd1').toString('base64'),
  },
};

const secret2 = {
  metadata: {
    namespace: 'bar',
    name: 'not-my-secret-name',
  },
  data: {
    username: Buffer.from('myuser2').toString('base64'),
    password: Buffer.from('mypwd2').toString('base64'),
  },
};

const secret3 = {
  metadata: {
    namespace: 'dummy',
    name: 'my-secret-name',
  },
  data: {
    username: Buffer.from('myuser3').toString('base64'),
    password: Buffer.from('mypwd3').toString('base64'),
  },
};

const bma2 = Object.assign({}, bma1, {
  status: {
    conditions: [
      {
        type: 'condType1',
        status: 'False',
        // the rest is unused
      },
      {
        type: 'condType2',
        status: 'True',
      },
      {
        type: 'condType3',
        status: 'False',
      },
      {
        type: 'condType4',
        status: 'True',
      },
    ],
  },
});


const bma3 = Object.assign({}, bma1, {
  status: {
    conditions: [
      {
        type: 'condType1',
        status: 'True',
        // the rest is unused
      },
      {
        type: 'condType2',
        status: 'True',
      },
      {
        type: 'condType3',
        status: 'True',
      },
      {
        type: 'condType4',
        status: 'True',
      },
    ],
  },
});

const bma4 = Object.assign({}, bma1, {
  status: {
    condidtons: [],
  },
});

describe('Bare Metal Asset model', () => {
  test('BMA transform function', () => {
    expect(transform(bma1)).toMatchSnapshot();
    expect(transform(bma1, [secret1])).toMatchSnapshot();
    expect(transform(bma2, [secret2, secret1, secret3])).toMatchSnapshot();
    expect(transform(bma3, [secret3, secret2, secret1])).toMatchSnapshot();
    expect(transform(bma4, [secret1, secret2, secret3])).toMatchSnapshot();
  });
});
