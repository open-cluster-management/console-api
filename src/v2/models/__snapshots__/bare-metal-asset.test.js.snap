// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bare Metal Asset model BMA transform function 1`] = `
Object {
  "bmc": Object {
    "address": "bmcAddr1",
    "credentialsName": "my-secret-name",
    "password": undefined,
    "username": undefined,
  },
  "bootMACAddress": "addr1",
  "clusterDeployment": "cluster",
  "hardwareProfile": "my-profile",
  "metadata": Object {
    "name": "foo",
    "namespace": "bar",
  },
  "role": "role",
  "status": "",
}
`;

exports[`Bare Metal Asset model BMA transform function 2`] = `
Object {
  "bmc": Object {
    "address": "bmcAddr1",
    "credentialsName": "my-secret-name",
    "password": "mypwd1",
    "username": "myuser1",
  },
  "bootMACAddress": "addr1",
  "clusterDeployment": "cluster",
  "hardwareProfile": "my-profile",
  "metadata": Object {
    "name": "foo",
    "namespace": "bar",
  },
  "role": "role",
  "status": "",
}
`;

exports[`Bare Metal Asset model BMA transform function 3`] = `
Object {
  "bmc": Object {
    "address": "bmcAddr1",
    "credentialsName": "my-secret-name",
    "password": "mypwd1",
    "username": "myuser1",
  },
  "bootMACAddress": "addr1",
  "clusterDeployment": "cluster",
  "hardwareProfile": "my-profile",
  "metadata": Object {
    "name": "foo",
    "namespace": "bar",
  },
  "role": "role",
  "status": "condType1",
}
`;

exports[`Bare Metal Asset model BMA transform function 4`] = `
Object {
  "bmc": Object {
    "address": "bmcAddr1",
    "credentialsName": "my-secret-name",
    "password": "mypwd1",
    "username": "myuser1",
  },
  "bootMACAddress": "addr1",
  "clusterDeployment": "cluster",
  "hardwareProfile": "my-profile",
  "metadata": Object {
    "name": "foo",
    "namespace": "bar",
  },
  "role": "role",
  "status": "Ready",
}
`;

exports[`Bare Metal Asset model BMA transform function 5`] = `
Object {
  "bmc": Object {
    "address": "bmcAddr1",
    "credentialsName": "my-secret-name",
    "password": "mypwd1",
    "username": "myuser1",
  },
  "bootMACAddress": "addr1",
  "clusterDeployment": "cluster",
  "hardwareProfile": "my-profile",
  "metadata": Object {
    "name": "foo",
    "namespace": "bar",
  },
  "role": "role",
  "status": "",
}
`;
